File: ios/Classes/FlutterStripeTerminal.swift
//
//  FlutterStripeTerminal.swift
//  flutter_stripe_terminal
//
//  Created by Vishal Dubey on 03/12/21.
//

import Foundation
import StripeTerminal
import Flutter

class FlutterStripeTerminal {
    static let shared = FlutterStripeTerminal()
    
    var serverUrl: String?
    var authToken: String?
    var requestType: String?
    var availableReaders: [Reader]?
    var discoverCancelable: Cancelable?
    
    
    func setConnectionTokenParams(serverUrl: String, authToken: String, requestType:String, result: FlutterResult) {
        self.serverUrl = serverUrl
        self.authToken = authToken
        self.requestType = requestType
        result(true)
    }
    
    func searchForReaders(simulated: Bool, result: @escaping FlutterResult) {
        let config = DiscoveryConfiguration(discoveryMethod: DiscoveryMethod.bluetoothScan, simulated: simulated)
        self.discoverCancelable = Terminal.shared.discoverReaders(config, delegate: FlutterStripeTerminalEventHandler.shared) { error in
            DispatchQueue.main.async {
                if let error = error {
                    result(error)
                } else {
                    result(true)
                }
            }
        }
    }
    
    func connectToReader(readerSerialNumber: String, locationId: String, result: @escaping FlutterResult) {
        let conn = Terminal.shared.connectedReader
        if conn == nil {
            let selectedReaders = self.availableReaders?.filter { reader in
                return reader.serialNumber == readerSerialNumber
            }
            if (!selectedReaders!.isEmpty) {
                Terminal.shared.connectBluetoothReader(selectedReaders![0], delegate: FlutterStripeTerminalEventHandler.shared, connectionConfig: BluetoothConnectionConfiguration(locationId: locationId, autoReconnectOnUnexpectedDisconnect: true, autoReconnectionDelegate: FlutterStripeTerminalEventHandler.shared)) { _, error in
                    DispatchQueue.main.async {
                        if let error = error {
                            result(error)
                        } else {
                            result(true)
                        }
                    }
                }
            }
        }
    }
    
    func updateReader(result: @escaping FlutterResult) {
        let conn = Terminal.shared.connectedReader
        if conn != nil {
            Terminal.shared.installAvailableUpdate()
        }
    }
    
    
    func connectionsStatus(result: @escaping FlutterResult) {
        
        let conn = Terminal.shared.connectionStatus
        switch conn.rawValue {
        case 0:
            result("not_connected")
        case 1:
            result("connected")
        case 2:
            result("connecting")
            
            
        default:
            result("not_connected")
        }
        result(conn.rawValue)
    }
    
    
    func disconnectReader() {
        Terminal.shared.disconnectReader {error in
            DispatchQueue.main.async {
                if let error = error {
                    print("Error disconnection")
                    print(error)
                } else {
                    print("disconnection complete")
                    
                }
            }
            
        }
        
    }
    
    
    func processPayment(clientSecret: String, result: @escaping FlutterResult) {
        let terminal = Terminal.shared
        //  discoverCancelable?.cancel()
        
        
        
        terminal.retrievePaymentIntent(clientSecret: clientSecret) { retrievedIntent, retrievedError in
            if let retrievedPaymentIntent = retrievedIntent {
                terminal.collectPaymentMethod(retrievedPaymentIntent) { processedIntent, processedError in
                    if let processedPaymentIntent = processedIntent {
                        terminal.processPayment(processedPaymentIntent) {finalIntent, finalError in
                            if let finalPaymentIntent = finalIntent {
                                switch finalPaymentIntent.status.rawValue{
                                    
                                    
                                case 5:
                                    result([
                                        "paymentIntentId": finalPaymentIntent.stripeId,
                                        "status": "SUCCEDED"
                                    ])
                                    
                                default:
                                    break
                                }
                                
                                
                                
                            } else if let finalError = finalError {
                                DispatchQueue.main.async {
                                    result(finalError)
                                }
                            }
                        }
                    } else if let processedError = processedError {
                        DispatchQueue.main.async {
                            print(processedError)
                            result(processedError)
                        }
                    }
                }
            } else if let retrievedError = retrievedError {
                DispatchQueue.main.async {
                    print(retrievedError)
                    result(retrievedError)
                }
            }
        }
    }
    
}

File: android/src/main/kotlin/in/agnostech/flutter_stripe_terminal/FlutterStripeTerminal.kt
package `in`.agnostech.flutter_stripe_terminal

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.gson.Gson
import com.stripe.stripeterminal.Terminal
import com.stripe.stripeterminal.external.callable.*
import com.stripe.stripeterminal.external.models.*
import com.stripe.stripeterminal.log.LogLevel
import io.flutter.plugin.common.MethodChannel

class FlutterStripeTerminal {
    companion object {
        lateinit var serverUrl: String
        lateinit var authToken: String

        // lateinit var context: Context
        var simulated: Boolean = false
        var availableReadersList: List<Reader>? = null
        var flutterStripeTerminalEventHandler: FlutterStripeTerminalEventHandler? = null
        var cancelDiscovery: Cancelable? = null;


        // Choose the level of messages that should be logged to your console
        val logLevel = LogLevel.NONE

        // Create your token provider.
        val tokenProvider = TokenProvider()


        fun setConnectionTokenParams(
            serverUrl: String,
            authToken: String,
            result: MethodChannel.Result
        ) {
            this.serverUrl = serverUrl
            this.authToken = authToken
            result.success(true)
        }

        fun disconnectReader(result: MethodChannel.Result) {

            val terminal = Terminal.getInstance()

            if (terminal.connectionStatus.toString() == "CONNECTED") {
                cancelDiscovery?.cancel(object : Callback {
                    override fun onFailure(e: TerminalException) {
                        Handler(Looper.getMainLooper()).post {
                        }
                    }

                    override fun onSuccess() {
                        Handler(Looper.getMainLooper()).post {
                        }
                    }

                })
                Terminal.getInstance().disconnectReader(object : Callback {
                    override fun onFailure(e: TerminalException) {
                        Handler(Looper.getMainLooper()).post {
                            result.error(e.errorCode.toLogString(), e.message, null)
                        }
                    }

                    override fun onSuccess() {
                        Terminal.getInstance().clearCachedCredentials()
                        Handler(Looper.getMainLooper()).post {
                            result.success(true)
                        }
                    }

                })
            }
        }

        fun searchForReaders(simulated: Boolean, result: MethodChannel.Result) {

            val config = DiscoveryConfiguration(
                //timeout = 60,
                discoveryMethod = DiscoveryMethod.BLUETOOTH_SCAN,
                isSimulated = simulated,
            )

            if (Terminal.isInitialized()) {

                val terminal = Terminal.getInstance()

                if (terminal.connectionStatus.toString() == "NOT_CONNECTED") {

                    cancelDiscovery = Terminal.getInstance().discoverReaders(
                        config,
                        flutterStripeTerminalEventHandler!!.getDiscoveryListener(),
                        object : Callback {
                            override fun onSuccess() {
                                Handler(Looper.getMainLooper()).post {
                                    result.success(true)
                                }
                            }

                            override fun onFailure(e: TerminalException) {
                                Handler(Looper.getMainLooper()).post {
                                    result.error(e.errorCode.toLogString(), e.message, null)
                                }
                            }
                        })
                }
            }
        }

        fun connectToReader(
            readerSerialNumber: String,
            locationId: String,
            result: MethodChannel.Result
        ) {
            val terminal = Terminal.getInstance()



            if (terminal.connectionStatus.toString() == "NOT_CONNECTED") {
                val reader = availableReadersList!!.filter {
                    it.serialNumber == readerSerialNumber
                }

                if (reader.isNotEmpty()) {
                    Log.d("STRIPE_RECONNECT",reader[0].toString())

                    terminal.connectBluetoothReader(reader[0],
                        ConnectionConfiguration.BluetoothConnectionConfiguration(
                            locationId,
                            true,
                            flutterStripeTerminalEventHandler!!
                        ),
                        flutterStripeTerminalEventHandler!!.getBluetoothReaderListener(),
                        object : ReaderCallback {
                            override fun onFailure(e: TerminalException) {
                                Handler(Looper.getMainLooper()).post {
                                    result.error(e.errorCode.toLogString(), e.message, null)
                                }
                            }

                            override fun onSuccess(reader: Reader) {
                                Handler(Looper.getMainLooper()).post {
                                    result.success(true)
                                }
                            }
                        })
                }
            }
        }


        fun connectionStatus(result: MethodChannel.Result) {
            val terminal = Terminal.getInstance()
            result.success(terminal.connectionStatus.toString())


        }

        fun updateReader(result: MethodChannel.Result) {
            val terminal = Terminal.getInstance()

            if (terminal.connectionStatus.toString() == "CONNECTED") {

                terminal.installAvailableUpdate()
                result.success(true)
            }
        }

        fun checkUpdateReader(result: MethodChannel.Result) {
            val terminal = Terminal.getInstance()
            if (terminal.connectionStatus.toString() == "CONNECTED") {

            }
        }

        fun processPayment(clientSecret: String, result: MethodChannel.Result) {
            val terminal = Terminal.getInstance()

            terminal.retrievePaymentIntent(clientSecret, object : PaymentIntentCallback {
                override fun onFailure(e: TerminalException) {
                    Handler(Looper.getMainLooper()).post {
                        result.error(e.errorCode.toLogString(), e.message, null)
                    }
                }

                override fun onSuccess(paymentIntent: PaymentIntent) {
                    Handler(Looper.getMainLooper()).post {

                        terminal.collectPaymentMethod(
                            paymentIntent,
                            object : PaymentIntentCallback {
                                override fun onFailure(e: TerminalException) {
                                    Handler(Looper.getMainLooper()).post {
                                        result.error(e.errorCode.toLogString(), e.message, null)
                                    }
                                }

                                override fun onSuccess(paymentIntent: PaymentIntent) {

                                    terminal.processPayment(
                                        paymentIntent,
                                        object : PaymentIntentCallback {
                                            override fun onFailure(e: TerminalException) {
                                                Handler(Looper.getMainLooper()).post {
                                                    result.error(
                                                        e.errorCode.toLogString(),
                                                        e.message,
                                                        null
                                                    )
                                                }
                                            }

                                            override fun onSuccess(paymentIntent: PaymentIntent) {

                                                Handler(Looper.getMainLooper()).post {

                                                    //result.success(paymentIntent)
                                                    result.success(
                                                        mapOf(
                                                            "paymentIntentId" to paymentIntent.id,
                                                            "PaymentIntentStatus" to paymentIntent.status.toString(),
                                                        )
                                                    )
                                                }
                                            }
                                        })
                                }
                            })
                    }
                }
            })
        }
    }
}
